File: scooter-control/src/Button/_ButtonInterceptors/_ButtonInterceptorBase.py
# Python Imports
from abc import ABC, abstractmethod

# Main
class ButtonInterceptorBase(ABC):
    def __init__(self, reverse_signal: bool=False) -> None:
        self._reverse_signal = reverse_signal

    @property
    def value(self) -> bool:
        return self.get_value()
    def get_value(self) -> bool:
        if self._reverse_signal:
            return not self._get_value()
        
        return self._get_value()
    
    @abstractmethod
    def _get_value() -> bool:
        return False

File: scooter-control/libs/scootercontrol/button/interceptors/ButtonInterceptorBase.cpp
// Class Definition
#include "ButtonInterceptorBase.hpp"

ButtonInterceptorBase::ButtonInterceptorBase(bool reverseSignal)
    : _reverseSignal(reverseSignal) {}

bool ButtonInterceptorBase::isPressed() {
    return _reverseSignal
        ? !_isPressed()
        : _isPressed();
}
